### PR TITLE
feat(core): store customer_acceptance in the payment_methods table

### PR DESCRIPTION
**Pull request by Aprabhat19**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

store customer_acceptance in the payment_methods table and some changes to setup_future_usage

- setup_future_usage should be inferred from the create and if passed in confirm should override the one in create
- If customer_acceptance is passed then only the card will be saved on hyper switch's end

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
 

-make a on_session payment without customer_acceptance, card will not be saved
-> Create
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_O7Znp5PhvGpSeE22Hxn4K8snHW6ObRfbA8R1isvLhUXpJRGHOINvlGufM4mGDAqj' \
--data-raw ' {"amount": 6677,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "authentication_type": "no_three_ds",
    "customer_id": "97cc",
    "email": "something@gmail.com",
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "metadata": {
        "order_details": {
            "product_name": "Apple iphone 15",
            "quantity": 1
        }
    },
    "business_country": "US",
    "business_label": "default",
"setup_future_usage": "on_session"
            
}'
```
-> Confirm
```
curl --location 'http://localhost:8080/payments/pay_IHI92Rlo3i6kwihECXVr/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_O7Znp5PhvGpSeE22Hxn4K8snHW6ObRfbA8R1isvLhUXpJRGHOINvlGufM4mGDAqj' \
--data '{
    "confirm": true,
      "payment_type": "normal",
    "payment_method": "card",

    "payment_method_type": "debit",
    "payment_method_data": {
        "card": {
            
      
            

            
       "card_number":"4242424242424242",
            "card_exp_month": "11",
            "card_exp_year": "2040",
            "card_holder_name": "AKA",
            "card_cvc": "123",
            "nick_name": "HELO"
        }
    }
}'
```
Screenshot 2024-03-05 at 1 47 56 PM

-make a off_session payment  without mandate_type ,i.e., MIT with customer_acceptance, the card will be saved on our end 
-> Create 
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_O7Znp5PhvGpSeE22Hxn4K8snHW6ObRfbA8R1isvLhUXpJRGHOINvlGufM4mGDAqj' \
--data-raw ' {"amount":0,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "authentication_type": "no_three_ds",
    "customer_id": "q7cc",
    "email": "something@gmail.com",
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "metadata": {
        "order_details": {
            "product_name": "Apple iphone 15",
            "quantity": 1
        }
    },
    "business_country": "US",
    "business_label": "default",
"setup_future_usage": "off_session"    
}'
```
-> Confirm

```
curl --location 'http://localhost:8080/payments/pay_IHI92Rlo3i6kwihECXVr/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_O7Znp5PhvGpSeE22Hxn4K8snHW6ObRfbA8R1isvLhUXpJRGHOINvlGufM4mGDAqj' \
--data '{
    "confirm": true,
      "payment_type": "setup_mandate",
        
      "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "in sit",
                "user_agent": "amet irure esse"
            }},

    "payment_method": "card",

    "payment_method_type": "debit",
    "payment_method_data": {
        "card": {
            
 
       "card_number":"4242424242424242",
            "card_exp_month": "11",
            "card_exp_year": "2040",
            "card_holder_name": "AKA",
            "card_cvc": "123",
            "nick_name": "HELO"
        }
    }
}'
```
Screenshot 2024-03-05 at 1 45 30 PM

Screenshot 2024-03-05 at 1 46 40 PM

-to save the card both, customer_acceptance and setup_future_usage must be passed

- if its a mandate_payment , we can have off_session in create with mandate type and customer_acceptance in confirm and also we can have customer_acceptance inside the mandate_data field which is for backwards compatibility currently there in mandate_data

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 1]
```diff
@@ -628,16 +632,16 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve> ValidateRequest<F, api::Paymen
...
+        // if request.mandate_data.is_none()
+        //     && request
+        //         .setup_future_usage
+        //         .map(|fut_usage| fut_usage == enums::FutureUsage::OffSession)
+        //         .unwrap_or(false)
+        // {
+        //     Err(report!(errors::ApiErrorResponse::PreconditionFailed {
+        //         message: "`setup_future_usage` cannot be `off_session` for normal payments".into()
+        //     }))?
+        // }
```
Comment (vspecky) [inline crates/router/src/core/payments/operations/payment_create.rs:644]: remove this

File: crates/router/src/core/payments/operations/payment_update.rs [hunk 2]
```diff
@@ -683,16 +684,16 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve> ValidateRequest<F, api::Paymen
...
+        // if request.mandate_data.is_none()
+        //     && request
+        //         .setup_future_usage
+        //         .map(|fut_usage| fut_usage == storage_enums::FutureUsage::OffSession)
+        //         .unwrap_or(false)
+        // {
+        //     Err(report!(errors::ApiErrorResponse::PreconditionFailed {
+        //         message: "`setup_future_usage` cannot be `off_session` for normal payments".into()
+        //     }))?
+        // }
```
Comment (vspecky) [inline crates/router/src/core/payments/operations/payment_update.rs:696]: remove this

File: crates/router/src/core/payments/tokenization.rs [hunk 3]
```diff
@@ -66,16 +68,16 @@ where
...
-                .unwrap_or(false);
+                .get_customer_acceptance()
+                .encode_to_value()
+                .change_context(errors::ApiErrorResponse::InternalServerError)
+                .attach_printable("Unable to serialize customer acceptance to value")?;
 
-            let pm_id = if future_usage_validation {
+            let pm_id = if resp.request.get_customer_acceptance().is_some()
+                && resp.request.get_setup_future_usage().is_some()
+            {
```
Comment (vspecky) [inline crates/router/src/core/payments/tokenization.rs:80]: Are we sure we don't need the `future_usage_validation` condition here?

**Review by vspecky — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/tokenization.rs (hunk 3 above)
Comment (Aprabhat19) [reply crates/router/src/core/payments/tokenization.rs:80]: Yes, we don't now that `customer_acceptance` is common between both of the `setup_future_usages` we don't want to say, that save the card only when `off_session+customer_acceptance` OR `on_session`.

Comment (vspecky) [reply crates/router/src/core/payments/tokenization.rs:80]: right

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 4]
```diff
@@ -281,6 +281,17 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve>
...
+
+        if Some(FutureUsage::OnSession) == request.setup_future_usage
+            && payment_attempt.mandate_details.is_some()
+        {
+            Err(error_stack::report!(
+                errors::ApiErrorResponse::PreconditionFailed {
+                    message: "`setup_future_usage` must be `off_session` for mandates".into()
+                }
+            ))?
+        };
```
[the remark this answers is not in the document]
Comment (Aprabhat19) [reply crates/router/src/core/payments/operations/payment_create.rs:293]: Yes , I have just added it in create , confirm update and complete authorize

**Review by jagan-jaya — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]
Dashboard specific changes look fine.

**Pull request merged**